### DISCUSSION TITLE
Edge classifier between feature patches is not converging.

### DISCUSSION THREAD

**Discussion by tails1234** [pyg-team/pytorch_geometric#7993] [Q&A] [answered]
Hi,

I have patches of features (14x8x8 each) from a CNN, each vertex is connected to all possibly related vertices/feature patches. I need to group related patches.
The dataset contains 12k graphs, each with possibly different number of vertices & edges. I made  `edge_index` as undirected with self loops.

I know that vertices that are related to each other would have a close values -values range from (-2,2)- in the feature patches, so I think this should be discriminative enough.

I used  three `EdgeConv` layers  followed by three fully connected layers to get a vertex embedding, then, for classification of edges I concatenate 2 vertex feature embeddings for each edge, followed by three fc layers as a classifier & softmax. BCE for loss is hovering at 50.

what could possibly be the issue here for the network not learning? I am not entirely sure if this could be done using gnns, that is why I am asking.

cheers

**Comment by rusty1s** [comment]
I personally think it is very important to incorporate relative spatial distance between patches as information to the GNN, e.g., this neighbor patch lays at the top/left of the center patch. Otherwise, you are basically learning on patches without any translation information, which makes this learning task pretty hard. Let me know if you are already doing this.

**Reply by tails1234** [reply]
Yes, I am already doing that. the 14th plane is the normalized  positional coords of each patch relative to the origin of the image, also, a one-hot-encoding to each type (out of 13 patch types).

**Comment by tails1234** [comment, marked as answer]
I think I found what I was doing wrong. The network was not learning because:
- softmax function was averaging over the 1st dimension instead of the 2nd dimension for a binary classification where the output is (edges, 2).
- the learning rate was too high (0.01) I changed it to be (lr = 5e-4).
- I normalized vertex input features & position coordinates.
now the loss starts at ~5.0 and decreased to 0.3 by epoch 6.
Thank you rusty1s. cheers =)
